handler/graphql.go

```diff
@@ -493,3 +519,92 @@ func sendError(w http.ResponseWriter, code int, errors ...*gqlerror.Error) {
 func sendErrorf(w http.ResponseWriter, code int, format string, args ...interface{}) {
 	sendError(w, code, &gqlerror.Error{Message: fmt.Sprintf(format, args...)})
 }
+
+func processMultipart(r *http.Request, request *params, uploadMaxMemory int64) error {
+	// Parse multipart form
+	if err := r.ParseMultipartForm(uploadMaxMemory); err != nil {
+		return errors.New("failed to parse multipart form")
+	}
+
+	// Unmarshal operations
+	var operations map[string]interface{}
+	if err := jsonDecode(strings.NewReader(r.Form.Get("operations")), &operations); err != nil {
+		return errors.New("operations form field could not be decoded")
+	}
+
+	// Unmarshal map
+	var uploadsMap = map[string][]string{}
+	if err := json.Unmarshal([]byte(r.Form.Get("map")), &uploadsMap); err != nil {
+		return errors.New("map form field could not be decoded")
+	}
+
+	// Read form files and add them to operations variables
+	var upload graphql.Upload
+	for key, path := range uploadsMap {
+		file, header, err := r.FormFile(key)
+		if err != nil {
+			return fmt.Errorf("failed to get key %s from form", key)
+		}
+		upload = graphql.Upload{
+			File:     file,
+			Size:     header.Size,
+			Filename: header.Filename,
+		}
+		if len(path) != 1 || !strings.HasPrefix(path[0], "variables.") {
+			return fmt.Errorf("invalid value for key %s", key)
+		}
+		err = addUploadToOperations(operations, upload, path[0])
+		if err != nil {
+			return err
+		}
+	}
+
+	// set request variables
+	if value, ok := operations["operationName"]; ok && value != nil {
+		request.OperationName = value.(string)
+	}
+	if value, ok := operations["query"]; ok && value != nil {
+		request.Query = value.(string)
+	}
+	if value, ok := operations["variables"]; ok && value != nil {
+		request.Variables = value.(map[string]interface{})
+	}
+
+	return nil
+}
+
+func addUploadToOperations(operations interface{}, upload graphql.Upload, path string) error {
+	var parts []interface{}
```

Comment:
This could be cleaned up a bit, the top loop isnt adding anything:
```go
func addUploadToOperations(request *params, upload graphql.Upload, path string) error {
	var ptr interface{} = request.Variables
	parts := strings.Split(path, ".")

	// skip the first part (variables) because we started there
	for i, p := range parts[1:] {
		last := i == len(parts)-2
		if ptr == nil {
			return fmt.Errorf("variables is missing, path: %s", path)
		}
		if index, parseNbrErr := strconv.Atoi(p) ; parseNbrErr == nil {
			if last {
				ptr.([]interface{})[index] = upload
			} else {
				ptr = ptr.([]interface{})[index]
			}
		} else {
			if last {
				ptr.(map[string]interface{})[p] = upload
			} else {
				ptr = ptr.(map[string]interface{})[p]
			}
		}
	}

	return nil
}
```

Reply:
Done